### PR TITLE
ROC-3054 Check country

### PR DESCRIPTION
**Pull request by robertparkinson**
### JIRA link
https://tools.hmcts.net/jira/browse/ROC-3054

### Change description
Check the country for the claimant and defendant and show an error if the country is not allowed

### Work checklist

- [ ] Unit tests added where applicable
- [ ] Route tests added for new pages
- [ ] New pages included in a11y tests
- [ ] Task list and task completeness checks updated
- [ ] Check and send page updated
- [ ] Routes, page content and page flows of new features are toggled 
- [ ] UI changes look good on mobile
- [ ] Required Google Analytics events are being sent 

### Developer self-QA run statement

- [x] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/main/app/common/country.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  static isDefendantCountry (value: string): boolean {
+    return Country.defendantCountries().filter(country => country.name.toLowerCase() === value.toLowerCase()).length > 0
+  }
+
+  static async getCountry (postcode: string): Promise<string> {
+    if (postcode === undefined || postcode.length === 0) {
+      return ''
+    } else {
+      let country = ''
+      await postcodeClient.lookupPostcode(postcode)
```
Comment (timja-kainos) [inline src/main/app/common/country.ts]: why are you looking this up again? isn't it already in the model?

File: src/main/app/common/country.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        .then((postcodeInfoResponse: PostcodeInfoResponse) => {
+          if (postcodeInfoResponse.valid) {
+            const object: PostcodeInfoCountry = postcodeInfoResponse.country as any as PostcodeInfoCountry
```
Comment (yarpo) [inline src/main/app/common/country.ts]: `postcodeInfoResponse.country as any as PostcodeInfoCountry`?

File: src/main/app/pdf/mappers/theirDetailsMapper.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ export class TheirDetailsMapper {
       address: {
         lineOne: party.address.line1,
         lineTwo: party.address.line2,
+        lineThree: party.address.line3,
```
Comment (yarpo) [inline src/main/app/pdf/mappers/theirDetailsMapper.ts]: why we keep doing it :D ?

File: src/test/app/common/country.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
+/* tslint:disable:no-unused-expression */
+
+import * as chai from 'chai'
+import * as chaiAsPromised from 'chai-as-promised'
+chai.use(chaiAsPromised)
+
+import { Country } from 'app/common/country'
+
+describe('Country', () => {
```
Comment (yarpo) [inline src/test/app/common/country.ts]: there is much more logic to cover in tests in `app/common/country` file.

File: src/main/app/common/country.ts (hunk 2 above)
Comment (robertparkinson) [reply src/main/app/common/country.ts]: the response from the postcodelookup client is a JSON object so when I try to parse it is throws an error saying it's already JSON. when I try to just assign it to the variable TypeScript wont compile it as doesn't see it as a proper casting to an object so had to cast it to any to then cast it as a PostcodeInfoCountry

File: src/main/app/common/country.ts (hunk 1 above)
Comment (robertparkinson) [reply src/main/app/common/country.ts]: The model doesn't save the returned object from the javascript call and it is possible that someone manually enters the address. we still need to do the check on those entries

File: src/test/app/common/country.ts (hunk 4 above)
Comment (robertparkinson) [reply src/test/app/common/country.ts]: I couldn't figure out how to mock returning Promises :(

File: src/main/app/pdf/mappers/theirDetailsMapper.ts (hunk 3 above)
Comment (robertparkinson) [reply src/main/app/pdf/mappers/theirDetailsMapper.ts]: doing what? I saw I had missed the line3 from a previous story so just added it in here :)

File: src/test/app/common/country.ts (hunk 4 above)
Comment (yarpo) [reply src/test/app/common/country.ts]: the way we do it in unit tests for routes wouldn't work here?

File: src/main/app/common/country.ts (hunk 1 above)
Comment (geek-soft) [reply src/main/app/common/country.ts]: Please do not use `then / catch` together with async / await.

File: src/main/app/common/country.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  static defendantCountries (): Country[] {
+    return [
+      Country.WALES,
+      Country.ENGLAND
+    ]
+  }
+
+  static async isValidClaimantAddress (form: Form<PartyDetails>): Promise<Form<PartyDetails>> {
```
Comment (geek-soft) [inline src/main/app/common/country.ts]: The method name is misleading. Methods starting with `is` are supposed to answer question in yes / no or true / false way. That method in turn populates validation errors on provided form.

File: src/main/app/common/country.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+          error.constraints = { amount: 'The country must be England or Wales' }
+          form.errors.push(new FormValidationError(error))
+        }
+      })
+    }
+
+    return form
+  }
+
+  static isClaimantCountry (value: string): boolean {
```
Comment (geek-soft) [inline src/main/app/common/country.ts]: The method name is inaccurate as well. I would read the method name in the way `is provided country a claimant one` which is not what this method is doing.

File: src/main/app/common/country.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
+import { Form, FormValidationError } from 'forms/form'
```
Comment (geek-soft) [inline src/main/app/common/country.ts]: This file has too many different constructs that fulfil so many different responsibilities. That gives the file too many reasons for change. Please split responsibilities in better way.

File: src/main/app/common/country.ts (hunk 5 above)
Comment (geek-soft) [inline src/main/app/common/country.ts]: This method is very similar to `isValidDefendantAddress` one. Please refactor code to remove code duplicates.

File: src/main/features/claim/routes/claimant-individual-details.ts [hunk 8]
```diff
@@ -26,18 +27,24 @@ export default express.Router()
...
     ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
-      const form: Form<IndividualDetails> = req.body
+      let form: Form<IndividualDetails> = req.body
+
       if (form.hasErrors()) {
         renderView(form, res)
       } else {
-        const draft: Draft<DraftClaim> = res.locals.claimDraft
-        const user: User = res.locals.user
+        form = await Country.isValidClaimantAddress(form)
```
Comment (geek-soft) [inline src/main/features/claim/routes/claimant-individual-details.ts]: That does not read well.

File: src/test/app/common/country.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    })
+
+    it('should return country for known country', () => {
+      Country.all().forEach(country => {
+        chai.expect(Country.valueOf(country.value)).to.be.equal(country)
+      })
+    })
+  })
+
+  describe('isClaimantCountry', () => {
```
Comment (geek-soft) [inline src/test/app/common/country.ts]: Please add missing tests cases for other methods you implemented in the `country.ts` file.

File: src/main/app/common/country.ts (hunk 6 above)
Comment (robertparkinson) [reply src/main/app/common/country.ts]: that is what this method is doing

File: src/main/app/common/country.ts (hunk 5 above)
Comment (robertparkinson) [reply src/main/app/common/country.ts]: will do

File: src/main/app/forms/validation/validators/checkCountry.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    if (!postcodeInfoResponse.valid) {
+      return true
+    }
+    const country = postcodeInfoResponse.country.name
+    const countries: Country[] = args.constraints[0]
+
+    return countries.filter(result => result.name.toLowerCase() === country.toLowerCase()).length > 0
+  }
+
+  defaultMessage (args: ValidationArguments) {
```
Comment (geek-soft) [inline src/main/app/forms/validation/validators/checkCountry.ts]: Please simplify by providing generic error message saying `Country is not permitted` or similar. Current implementation strongly tights the default message that should be generic with specific validation rules for claimant / defendant country. That is not desirable. When default message is inappropriate then it can be overridden.

File: src/main/features/claim/helpers/eligibility/validationGroups.ts [hunk 11]
```diff
@@ -9,4 +9,6 @@ export class ValidationGroups {
   static readonly DEFENDANT_ADDRESS: string = 'defendant-address'
   static readonly GOVERNMENT_DEPARTMENT: string = 'government-department'
   static readonly CLAIM_IS_FOR_TENANCY_DEPOSIT: string = 'claim-is-for-tenancy-deposit'
+  static readonly DEFENDANT: string = 'defendant'
```
Comment (geek-soft) [inline src/main/features/claim/helpers/eligibility/validationGroups.ts]: I would either revert the change or put these constants in different place. Eligibility validation groups does not look as the right place for it.

File: src/main/features/claim/routes/claimant-individual-details.ts [hunk 12]
```diff
@@ -24,9 +25,10 @@ export default express.Router()
   })
   .post(
     Paths.claimantIndividualDetailsPage.uri,
-    FormValidator.requestHandler(IndividualDetails, IndividualDetails.fromObject),
+    FormValidator.requestHandler(IndividualDetails, IndividualDetails.fromObject, ValidationGroups.CLAIMANT),
     ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
-      const form: Form<IndividualDetails> = req.body
+      let form: Form<IndividualDetails> = req.body
```
Comment (geek-soft) [inline src/main/features/claim/routes/claimant-individual-details.ts]: Please revert, no need for that change.

File: src/main/features/claim/routes/claimant-organisation-details.ts [hunk 13]
```diff
@@ -25,9 +26,10 @@ export default express.Router()
   })
   .post(
     Paths.claimantOrganisationDetailsPage.uri,
-    FormValidator.requestHandler(OrganisationDetails, OrganisationDetails.fromObject, 'claimant'),
+    FormValidator.requestHandler(OrganisationDetails, OrganisationDetails.fromObject, ValidationGroups.CLAIMANT),
     ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
-      const form: Form<OrganisationDetails> = req.body
+      let form: Form<OrganisationDetails> = req.body
```
Comment (geek-soft) [inline src/main/features/claim/routes/claimant-organisation-details.ts]: Please revert, no need for that change.

File: src/main/features/claim/routes/claimant-sole-trader-details.ts [hunk 14]
```diff
@@ -25,9 +26,10 @@ export default express.Router()
   })
   .post(
     Paths.claimantSoleTraderOrSelfEmployedDetailsPage.uri,
-    FormValidator.requestHandler(SoleTraderDetails, SoleTraderDetails.fromObject),
+    FormValidator.requestHandler(SoleTraderDetails, SoleTraderDetails.fromObject, ValidationGroups.CLAIMANT),
     ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
-      const form: Form<SoleTraderDetails> = req.body
+      let form: Form<SoleTraderDetails> = req.body
```
Comment (geek-soft) [inline src/main/features/claim/routes/claimant-sole-trader-details.ts]: Please revert, no need for that change.

File: src/main/features/claim/routes/defendant-individual-details.ts [hunk 15]
```diff
@@ -24,9 +25,10 @@ export default express.Router()
   })
   .post(
     Paths.defendantIndividualDetailsPage.uri,
-    FormValidator.requestHandler(IndividualDetails, IndividualDetails.fromObject),
+    FormValidator.requestHandler(IndividualDetails, IndividualDetails.fromObject, ValidationGroups.DEFENDANT),
     ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
-      const form: Form<IndividualDetails> = req.body
+      let form: Form<IndividualDetails> = req.body
```
Comment (geek-soft) [inline src/main/features/claim/routes/defendant-individual-details.ts]: Please revert, no need for that change.

File: src/main/features/claim/routes/defendant-organisation-details.ts [hunk 16]
```diff
@@ -25,9 +26,10 @@ export default express.Router()
   })
   .post(
     Paths.defendantOrganisationDetailsPage.uri,
-    FormValidator.requestHandler(OrganisationDetails, OrganisationDetails.fromObject, 'defendant'),
+    FormValidator.requestHandler(OrganisationDetails, OrganisationDetails.fromObject, ValidationGroups.DEFENDANT),
     ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
-      const form: Form<OrganisationDetails> = req.body
+      let form: Form<OrganisationDetails> = req.body
```
Comment (geek-soft) [inline src/main/features/claim/routes/defendant-organisation-details.ts]: Please revert, no need for that change.

File: src/main/features/claim/routes/defendant-sole-trader-details.ts [hunk 17]
```diff
@@ -25,9 +26,10 @@ export default express.Router()
   })
   .post(
     Paths.defendantSoleTraderOrSelfEmployedDetailsPage.uri,
-    FormValidator.requestHandler(SoleTraderDetails, SoleTraderDetails.fromObject),
+    FormValidator.requestHandler(SoleTraderDetails, SoleTraderDetails.fromObject, ValidationGroups.DEFENDANT),
     ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
-      const form: Form<SoleTraderDetails> = req.body
+      let form: Form<SoleTraderDetails> = req.body
```
Comment (geek-soft) [inline src/main/features/claim/routes/defendant-sole-trader-details.ts]: Please revert, no need for that change.

File: src/main/features/response/routes/your-details.ts [hunk 18]
```diff
@@ -76,7 +76,7 @@ export default express.Router()
     Paths.defendantYourDetailsPage.uri,
     FormValidator.requestHandler(PartyDetails, deserializeFn, 'response'),
     ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction): Promise<void> => {
-      const form: Form<PartyDetails> = req.body
+      let form: Form<PartyDetails> = req.body
```
Comment (geek-soft) [inline src/main/features/response/routes/your-details.ts]: Please revert, no need for that change.

File: src/main/app/forms/validation/validators/checkCountry.ts [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    } else if (countries.length === 2) {
+      return AddressValidationErrors.DEFENDANT_COUNTRY_REQUIRED
+    } else {
+      return AddressValidationErrors.POSTCODE_REQUIRED
+    }
+  }
+}
+
+/**
+ * Verify is a within age limit.
```
Comment (geek-soft) [inline src/main/app/forms/validation/validators/checkCountry.ts]: Please update.

File: src/main/app/forms/validation/validators/checkCountry.ts [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+ */
+export function CheckCountry (countries: Country[], validationOptions?: ValidationOptions) {
```
Comment (geek-soft) [inline src/main/app/forms/validation/validators/checkCountry.ts]: I would recommend changing name to fit into the existing convention. What about 'IsPostcodeInPermittedCountry' or 'IsCountrySupported'? Or something along these lines?

File: src/main/app/forms/models/address.ts [hunk 21]
```diff
@@ -62,8 +67,10 @@ export class Address implements CompletableTask {
   city?: string
 
   @ValidateIf(o => o.addressVisible, { groups: ['claimant', 'defendant', 'response'] })
-  @IsDefined({ message: ValidationErrors.POSTCODE_REQUIRED, groups: ['claimant', 'defendant', 'response'] })
-  @IsNotBlank({ message: ValidationErrors.POSTCODE_REQUIRED, groups: ['claimant', 'defendant', 'response'] })
+  @IsDefined({ message: ValidationErrors.POSTCODE_REQUIRED, groups: ['response'] })
+  @IsNotBlank({ message: ValidationErrors.POSTCODE_REQUIRED, groups: ['response'] })
+  @CheckCountry(Country.all(), { groups: ['claimant'] })
+  @CheckCountry(Country.defendantCountries(), { groups: ['defendant'] })
```
Comment (geek-soft) [inline src/main/app/forms/models/address.ts]: **Business question:** Should we check defendant address when defendant confirms their details during the response time? What if claimant provided defendant address within allowed range however when defendant tries to respond it happens that his address is outside the allowed range?

Comment (robertparkinson) [reply src/main/app/forms/models/address.ts]: I asked art this and he said it only matters when issuing the claim. If the defendant updates their address during the response we don't need to check it

File: src/main/features/claim/routes/defendant-company-details.ts [hunk 22]
```diff
@@ -24,7 +24,7 @@ export default express.Router()
   })
   .post(
     Paths.defendantCompanyDetailsPage.uri,
-    FormValidator.requestHandler(CompanyDetails, CompanyDetails.fromObject, 'defendant'),
+    FormValidator.requestHandler(CompanyDetails, CompanyDetails.fromObject , 'defendant'),
```
Comment (geek-soft) [inline src/main/features/claim/routes/defendant-company-details.ts]: That change is not needed.

File: src/main/app/forms/validation/validators/isCountrySupported.ts [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+import { PostcodeInfoClient, PostcodeInfoResponse } from '@hmcts/postcodeinfo-client'
+import * as config from 'config'
+import { request } from 'client/request'
+
+const postcodeClient = new PostcodeInfoClient(config.get<string>('postcodeLookup.apiKey'), request)
+
+@ValidatorConstraint()
+export class CheckCountryConstraint implements ValidatorConstraintInterface {
+
+  async validate (value: any | string, args?: ValidationArguments): Promise<boolean> {
```
Comment (geek-soft) [inline src/main/app/forms/validation/validators/isCountrySupported.ts]: Please add unit tests for this validate method that tests the following:
- undefined value
- null value
- empty value
- invalid postcode
- not support postcode 
- supported postcode

Comment (robertparkinson) [reply src/main/app/forms/validation/validators/isCountrySupported.ts]: I've added the tests now

File: src/main/features/claim/routes/defendant-company-details.ts (hunk 22 above)
Comment (robertparkinson) [reply src/main/features/claim/routes/defendant-company-details.ts]: removed

File: src/test/app/forms/validation/validators/isCountrySupported.ts [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+import { Country } from 'app/common/country'
+import { ValidationArguments } from 'class-validator'
+
+describe('IsCountrySupported', () => {
+  const constraint: CheckCountryConstraint = new CheckCountryConstraint()
+
+  describe('validate', () => {
+
+    describe('should return true when ', () => {
+      it('given am undefined value', async () => {
```
Comment (geek-soft) [inline src/test/app/forms/validation/validators/isCountrySupported.ts]: Spelling, should be 'an' ;)

File: src/test/app/forms/validation/validators/isCountrySupported.ts [hunk 25]
```diff
@@ -0,0 +1,53 @@
+import { expect } from 'chai'
+import { CheckCountryConstraint } from 'forms/validation/validators/isCountrySupported'
+import { Country } from 'app/common/country'
+import { ValidationArguments } from 'class-validator'
+
+describe('IsCountrySupported', () => {
+  const constraint: CheckCountryConstraint = new CheckCountryConstraint()
+
+  describe('validate', () => {
```
Comment (geek-soft) [inline src/test/app/forms/validation/validators/isCountrySupported.ts]: That tests should mock responses from the postcode lookup API to not use the real one that comes with an difficulty of possessing working API token. Please use the same approach as we have in route tests that uses 'nock' http server.

File: src/test/app/forms/validation/validators/isCountrySupported.ts [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+          .get(/\/addresses\/\?postcode=.+/)
+          .reply(200, [])
+
+        nock(mockPostcode)
+          .get(/\/postcodes\/.+/)
+          .reply(200, mockInvalidPostcodeLookupResponse)
+        expect(await constraint.validate('SA6 7JL', validationArgs(Country.all()))).to.equal(true)
+      })
+
+      it('given an invalid postcode', async () => {
```
Comment (geek-soft) [inline src/test/app/forms/validation/validators/isCountrySupported.ts]: Despite the fact you use two different postcodes this test case is similar to the one above. Since you mock the response it does not really matter what postcode you pass, what matters is the response HTTP server will return.

Please remove one of these.

File: src/main/app/forms/validation/validators/isCountrySupported.ts [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+    if (value === undefined || value === null || value === '') {
+      return true
+    }
+    const postcodeInfoResponse: PostcodeInfoResponse = await postcodeClient.lookupPostcode(value)
```
Comment (geek-soft) [inline src/main/app/forms/validation/validators/isCountrySupported.ts]: Plase also add scenario where postcode lookup API returns 500 response to see how validator would behave in such scenario.

File: src/main/app/forms/validation/validators/isCountrySupported.ts [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+      const postcodeInfoResponse: PostcodeInfoResponse = await postcodeClient.lookupPostcode(value)
+      if (!postcodeInfoResponse.valid) {
+        return true
+      }
+      const country = postcodeInfoResponse.country.name
+      const countries: Country[] = args.constraints[0]
+
+      return countries.filter(result => result.name.toLowerCase() === country.toLowerCase()).length > 0
+    } catch (err) {
+      return true
```
Comment (geek-soft) [inline src/main/app/forms/validation/validators/isCountrySupported.ts]: We would like to print errors in the logs so that we can be aware there was a problem with postcode lookup. Please log an error.

File: src/main/app/forms/validation/validators/isCountrySupported.ts (hunk 23 above)
Comment (yarpo) [reply src/main/app/forms/validation/validators/isCountrySupported.ts]: @geek-soft do we want to test `null`?

File: src/main/app/pdf/mappers/theirDetailsMapper.ts (hunk 3 above)
Comment (yarpo) [reply src/main/app/pdf/mappers/theirDetailsMapper.ts]: No, I mean "lineThree" and "line3".

I though we had unify it.
